Update client guide for Python (#356)
algorithmiaio/dev-center: _pages/clients/python.md

---
layout: article

> title: "Python Client Guide"

Reviewer: Alright if we keep this as it is? Adding `Client Guide` is redundant when seeing the page in the side nav:

<img width="277" alt="Screen Shot 2020-11-04 at 11 18 35 AM" src="https://user-images.githubusercontent.com/8980868/98158903-89259e80-1e90-11eb-8fcc-65281d8cd139.png">

cc @JamesHoover

Reviewer 2: Yeah thats fine to remove

{% include video-responsive.html height="560" width="315" url="https://www.youtube.com/embed/bZB2vu0v6A0" %}

The Algorithmia Python client provides a native Python interface to the Algorithmia API, letting developers manage and call algorithms, work with data in object stores using the Algorithmia Data Portal, and access other features of the Algorithmia platform.

> This guide will cover setting up the client, calling an algorithm using direct user input, calling an algorithm that accesses data through the Algorithmia Data Portal, and using Algorithmia's Hosted Data service. For complete details about the Algorithmia API, please refer to the [API Docs](/developers/api/?python#introduction).

Reviewer 3: @JamesHoover I just noticed that the docs refer to data sources as "the data portal" which seems outdated since on the UI it says "Data Sources" https://algorithmia.com/data are we consistently calling /data page "Data Sources" or are we still going to brand it as "Data Portal"?

Author: My understanding is that the "Data Portal" is a platform feature which allows you to access multiple object stores (Hosted Data, S3, etc.) using the same code and URIs, while "Data Sources" are the object stores which the Data Portal has been configured to use.

Reviewer 3: Yeah, data portal has been used in the past, but since it's not in the UI anywhere (anymore), so it might be confusing to users if it only appears in the docs - gonna check with Hoover on this real quick

Reviewer 3: Got confirmation to remove the terminology "Data Portal" and we can just use "Data Sources" since that's the header of the page that it links to

Alternately, you can ensure that each of your servers interacting with your Algorithmia Enterprise instance have an environment variable named `ALGORITHMIA_API` and the client will use it.  The fallback API endpoint is always the hosted Algorithmia marketplace service at [https://api.algorithmia.com/](https://api.algorithmia.com/)
{% endif %}

Calling an Algorithm

> --------------------

Reviewer: nit: let's use the same header syntax as the rest of the file so it's consistent: `## Calling an Algorithm`

The first algorithm we'll call is a demo version of the algorithm used in the Algorithm Development [Getting Started](/developers/algorithm-development/your-first-algo) guide, which is available at [demo/Hello](/algorithms/demo/Hello). Looking at the [algorithm's documentation](/algorithms/demo/Hello/docs), it takes a string as input and returns a string.

In order to call an Algorithm from Python, we need to first create an algorithm object. With the client already instantiated, we can run the following code to create an object:

{% highlight python %}
algo = client.algo('demo/Hello') 
{% endhighlight %}

Then, we can use the `.pipe()` method to call the algorithm. We'll provide out input as the argument to the function, and then print the output using the `result` attribute:

{% highlight python %}
response = algo.pipe("Mr. Bond")
print(reponse.result)
{% endhighlight %}

Which should print the phrase, `Hello Mr. Bond`

### JSON and Python

> The Python client provides some ease of use abstractions for working with alrogithms with JSON inputs and outputs. When passing a Python array or dict into the `.pipe()` function, the library will automatically serialize it to JSON. For algorithms that return a JSON type, the `result` field of the response will contain an array or dict, as appropriate.

Reviewer: ```suggestion
The Python client provides some ease of use abstractions for working with algorithms with JSON inputs and outputs. When passing a Python array or dict into the `.pipe()` function, the library will automatically serialize it to JSON. For algorithms that return a JSON type, the `result` field of the response will contain an array or dict, as appropriate.
```

> The Python client provides some ease of use abstractions for working with alrogithms with JSON inputs and outputs. When passing a Python array or dict into the `.pipe()` function, the library will automatically serialize it to JSON. For algorithms that return a JSON type, the `result` field of the response will contain an array or dict, as appropriate.

Reviewer 3: ```suggestion
The Python client provides some ease of use abstractions for working with alrogithms with JSON inputs and outputs. When passing a Python array or dict into the `.pipe()` function, the library will automatically serialize it to JSON. Algorithms will return a JSON type and the `result` field of the response will contain an array or dict, as appropriate.
```

Reviewer 3: Since all algos return some sort of JSON response I changed the wording

> Let's look at an example using JSON and the [WebPredict/ListAnagrams](/algorithms/WebPredict/ListAnagrams) algorithm. The [algorithm docs](https://algorithmia.com/algorithms/WebPredict/ListAnagrams/docs) tell us that the algorithm takes an array of strings and filters it to remove any words that do not have an anagram in the array. We'll call this algorithm using an array of strings as the input:

Reviewer: Any URLs that refer to specific algorithms should be absolute. (If someone happens to be viewing the documentation in an enterprise setting, the relative path will attempt to load an algorithm that doesn't exist.)

```suggestion
Let's look at an example using JSON and the [WebPredict/ListAnagrams](https://algorithmia.com/algorithms/WebPredict/ListAnagrams) algorithm. The [algorithm docs](https://algorithmia.com/algorithms/WebPredict/ListAnagrams/docs) tell us that the algorithm takes an array of strings and filters it to remove any words that do not have an anagram in the array. We'll call this algorithm using an array of strings as the input:
```

Reviewer 3: And actually let's not use any algos that are user generated - we can use any from NLP user, demo user, web user,  Apache NLP user for instance https://algorithmia.com/tags/apache-opennlp

Reviewer 3: We can use [LDA](https://algorithmia.com/algorithms/nlp/LDA/docs) which takes multiple input types such as a file (for working with file section) or text for this section

{% highlight python %}
algo.set_options(timeout=60, stdout=False)
{% endhighlight %}

You can find more details in [API Docs](/developers/api/?python) > [Invoke an Algorithm](/developers/api/?python#invoke-an-algorithm).

### Error Handling

To be able to better develop across languages, Algorithmia has created a set of standardized errors that can be returned by either the platform or by the algorithm being run. In Python, API errors and Algorithm exceptions will result in calls to pipe throwing an `AlgoException`:

{% highlight python %}
client.algo('util/whoopsWrongAlgo').pipe('Hello, world!')
# Algorithmia.algo_response.AlgoException: algorithm algo://util/whoopsWrongAlgo not found
{% endhighlight %}

You can read more about [Error Handling](/developers/algorithm-development/algorithm-errors) in the [Algorithm Development](/developers/algorithm-development) section of the dev center.

Working with the Algorithmia Data Portal

> ----------------------------------------

Reviewer: nit: update to `## Working with the Algorithmia Data Portal` to match the syntax in the rest of the file.

For some algorithms, passing input to the algorithm at request time is sufficient, while others might have larger data requirements or need to preserve state between calls. Application developers can use Algorithmia's [Data Portal](/data) to store data as text, JSON, or binary, and access it via the Algorithmia [Data API](/developers/api/?python#data).

The Data API defines [connectors](/developers/api/?python#connectors) to a variety of storage providers, including Algorithmia [Hosted Data](/developers/data/hosted), Amazon S3, Google Cloud Storage, Azure Storage Blobs, and Dropbox. After creating a connection in the Data Portal, you can use the API to create, update, and delete directories and files and manage permissions across providers by making use of [Data URIs](/developers/api/#data-uris) in your code.

> In this example, we'll upload a file to Algorithmia's Hosted Data storage provider, invoke the "*TODO*" algorithm which will process the file, upload the result, and return a data URI which we can use to download the file.

Reviewer 3: ```suggestion
In this example, we'll upload a file to Algorithmia's [Hosted Data](/developers/data/hosted) storage provider, invoke the "*TODO*" algorithm which will process the file, upload the result, and return a data URI which we can use to download the file.
```

### Create a Data Collection

The documentation for "*TODO*" says that it takes a Data URI as input, but we still need a location to store the data. We'll create a new directory, then update its permissions so that it's publicly accessible:

{% highlight python %}
from Algorithmia.acl import ReadAcl, AclType
# Instantiate a DataDirectory object, set your data URI and call create
img_directory = client.dir("data://YOUR_USERNAME/img_directory")
# Create your data collection if it does not exist
if img_directory.exists() is False:
    img_directory.create()
# Change permissions on your data collection to public
img_directory.update_permissions(ReadAcl.public)
{% endhighlight %}

Instead of your username you can also use '.my' when calling algorithms. For more information about the '.my' pseudonym check out the [Hosted Data Guide]({{site.baseurl}}/data/hosted).
{: .notice-info}

### Upload Data to your Data Collection

> Now we're ready to upload an image file for processing. We'll use *TODO* something.jpg. First create a variable that holds the path in your data collection to the image file you will be uploading:

Reviewer 3: Not sure if *TODO* is a placeholder, but let's use a real algorithm for users to call (I can help find a good one, but having issues with username links working rn)

### Upload Data to your Data Collection

Now we're ready to upload an image file for processing. We'll use *TODO* something.jpg. First create a variable that holds the path in your data collection to the image file you will be uploading:

{% highlight python %}
# Download contents of file as a string
if client.file(text_file).exists() is True:
    input = client.file(text_file).getString()
img_file = "data://YOUR_USERNAME/img_directory/something.jpg"
{% endhighlight %}

This will get your file as a string, saving it to the variable `input`.  If the file was binary (an image, etc), you could instead use the function `.getBytes()` to retrieve the file's content as a byte array. For more image-manipulation tutorials, see the [Computer Vision Recipes]({{site.baseurl}}/tutorials/recipes/#computer-vision).

Now you've seen how to upload a local data file, check if a file exists in a data collection, and download the file contents.

For more methods on how to get a file using the Data API from a data collection go to the [API Specification](http://docs.algorithmia.com/#getting-a-file).

> {% if site.enterprise %}

Reviewer 3: Insights section looks like it was removed in new version - this section needs to appear in the new version as well w/enterprise flag

try:
    # Get the summary result of your file's contents
    print(algo.pipe(input).result)
    response = algo.pipe(img_file)
except Exception as error:
    # Algorithm error if, for example, the input is not correctly formatted
    print(error)
{% endhighlight %}

This guide used the the first chapter of [Jack London's Burning Daylight](https://en.wikisource.org/wiki/Burning_Daylight) and the Summarizer algorithm outputs:

"It was a quiet night in the Shovel. The miners were in from Moseyed Creek and the other diggings to the west, the summer washing had been good, and the men's pouches were heavy with dust and nuggets. MacDonald grinned and nodded, and opened his mouth to speak, when the front door swung wide and a man appeared in the light."

If you are interested in learning more about working with unstructured text data check out our guide [Introduction to Natural Language Processing](https://algorithmia.com/blog/introduction-natural-language-processing-nlp).
Once the algorithm has completed, `response.result` will contain the URI for the resulting file, which you can then download (or provide as input to another algorithm in a pipeline).

> ### Limits

Reviewer 3: This shouldn't be removed - didn't see it in the new version unless i skipped it

> If the file was text (an image, etc), you could instead use the function `.getString()` to retrieve the file's content as a string. For more methods on how to get a file using the Data API from a data collection go to the [API Specification](http://docs.algorithmia.com/#getting-a-file).

Reviewer: nit: can prevent an unnecessary redirect here by using the algorithmia.com domain:
```suggestion
If the file was text (an image, etc), you could instead use the function `.getString()` to retrieve the file's content as a string. For more methods on how to get a file using the Data API from a data collection go to the [API Specification](/developers/api/#get-a-file-or-directory).
```

> ------------------------

Reviewer: `## Additional Functionality`

> In addition to the functionality covered in this guide, the Python Client Library provides a complete interface to the Algorithmia platform, including managing Algorithms (and their versions and builds), organizations, users, source control, and the additional functionality of the Data API. You can find all the details, including Python examples, in the API Docs.

Reviewer 3: Add links for API Docs, Data API, organizations, users, and source control

Author: Where there were relevant Dev Center links, I added those, otherwise I used API Docs links. Would all API Docs links be preferred?

Reviewer 3: I just mean in this specific section to add https://algorithmia.com/developers/platform/organizations to "organizations", "source control" https://algorithmia.com/developers/algorithm-development/source-code-management, "API docs" https://algorithmia.com/developers/api/#introduction, "managing algorithms" https://algorithmia.com/developers/algorithm-development/algorithm-management more for a CTA then anything else - also let's remove "users" from that list because it's not really a page or section

Author: This was intended to reference the sections from the API Docs to provide a complete list of what you can do with the client, rather than links to more Dev Center pages. Is that not needed?

Author: Removed the API docs-specific links and replaced them with dev center links.

> ----------

Reviewer: `## Next Steps`

> If you're a developer looking to call algorithms from the marketplace or those created by your team members, you're ready to start coding! If you're a data scientist or developer who will be building and deploying new algorithms, you can move on to the [Algorithm Development > Getting Started](/developers/algorithm-development/your-first-algo/) guide.

Reviewer 3: ```suggestion
If you're a data scientist or developer who will be building and deploying new algorithms, you can move on to the [Algorithm Development > Getting Started](/developers/algorithm-development/your-first-algo/) guide. 
```

Reviewer 3: Just removing mention of marketplace, but I'll think of some better messaging here

This endpoint will replace a file if it already exists. If you wish to avoid replacing a file, check if the file exists before using this endpoint.
{: .notice-warning}

Confirm that the file was created by navigating to Algorithmia's [Hosted Data Source](/data/hosted) and finding your data collection and file.

You can also upload your data through the UI on Algorithmia's [Hosted Data Source](/data/hosted). For instructions on how to do this go to the [Hosted Data Guide]({{site.baseurl}}/data/hosted).

### Call the Algorithm

> Once the file has been uploaded, you are ready to call the algorithm. Create the algorithm object, then pass the required inputs. an image URI (which is stored in `img_file` in the code above), a URI for the image output, and the size of the pixels to be used, to `algo.pipe()`:

Reviewer 3: Just needs a capital A in "...required inputs. **an** image URI..."

{: .notice-info}

### Download the resulting file

The URI included in the algorithm output uses the `.algo` shortcut, so we'll need to modify it slighly to download the file by adding the algorithm name and author:

{% highlight python %}
output_uri = response.result['images'][0]['output']
download_uri = output_uri[:13] + `dlib/FaceDetection` + output_uri[13:]
{% endhighlight %}

Verify that the file that you want to download exists, and try downloading it to a new local file:

{% highlight python %}
# Download the file
if client.file(download_uri).exists() is True:
    local_file = client.file(download_uri).getFile()
{% endhighlight %}

This copies the file from your data collection and saves it as a file on your local machine, storing the filename in the variable `local_file`.

Alternately, if you just need the binary content of the file to be stored in a variable, you can retrieve the remote file's content without saving the actual file:

{% highlight python %}
# Download contents of file as a string
if client.file(download_uri).exists() is True:
    output = client.file(download_uri).getBytes()
{% endhighlight %}

> This will get the image as binary data, saving it to the variable `input`, which might be useful when writing algorithms that are part of an image processing pipeline. For more image-manipulation tutorials, see the [Computer Vision Recipes]({{site.baseurl}}/tutorials/recipes/#computer-vision). 

Reviewer 3: I think you meant to say "...the variable `output`..." instead of input

Author: I did mean input, because I was thinking about using this in the context of a pipeline or something where it was the input to the next algorithm. Maybe `image_data` would be better?
